File: test_quality_breakdown.py
#!/usr/bin/env python3
"""
Quality Component Breakdown for Divergent Measurement

Processes the first divergent measurement (#1) and extracts
detailed breakdown of each quality component
"""

import csv
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

# Add paths
sys.path.insert(0, str(Path(__file__).parent / "python_lib" / "src"))
sys.path.insert(0, str(Path(__file__).parent / "be_implementation_service" / "src"))

from weight_processor_lib.core.database.memory_store import InMemoryStore
from weight_processor_lib.core.processing.processor import Processor


def parse_timestamp(date_str: str) -> datetime:
    """Parse various timestamp formats"""
    if not date_str:
        return datetime.now(timezone.utc)

    try:
        if "T" in date_str:
            if date_str.endswith("Z"):
                return datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            else:
                return datetime.fromisoformat(date_str)
        elif " " in date_str:
            return datetime.fromisoformat(date_str.replace(" ", "T") + "+00:00")
        else:
            return datetime.fromisoformat(date_str + "T00:00:00+00:00")
    except Exception:
        return datetime.now(timezone.utc)


def get_default_config():
    """Get default configuration"""
    return {
        "database": {"backend": "memory"},
        "kalman": {
            "initial_variance": 0.364,
            "transition_covariance_weight": 0.018,
            "transition_covariance_trend": 0.00015,
            "observation_covariance": 3.49,
        },
        "quality_scoring": {
            "threshold": 0.5,
            "components": {
                "kalman_fit": {"weight": 0.3, "enabled": True},
                "temporal_consistency": {"weight": 0.25, "enabled": True},
                "anomaly_detection": {"weight": 0.25, "enabled": True},
                "source_reliability": {"weight": 0.1, "enabled": True},
                "trend_alignment": {"weight": 0.1, "enabled": True},
            },
        },
        "processing": {"enable_validation": True, "enable_quality_scoring": True},
        "reset": {"time_gap_days": 30, "weight_change_threshold_kg": 10},
        "snapshot": {"interval_hours": 24, "periodic_enabled": True},
        "adaptive_noise": {"enabled": True},
        "replay": {
            "buffered_replay_enabled": False,
            "buffer_hours": 24,
            "max_buffer_measurements": 100,
        },
    }


def main():
    csv_path = "test_user.csv"
    user_id = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7"
    target_measurement_id = "0bb4ca6c-d123-4461-8cae-a40297230843"  # Divergent measurement #1

    # Load CSV
    measurements = []
    with open(csv_path, "r") as f:
        reader = csv.DictReader(f)
        for row in reader:
            measurement_id = row.get("id") or row.get("measurement_id")
            if not measurement_id or row.get("user_id") != user_id:
                continue

            weight_str = (row.get("value_quantity") or row.get("weight") or "").strip()
            if not weight_str or weight_str.upper() == "NULL":
                continue

            try:
                weight = float(weight_str)
                if weight <= 0 or weight > 1000:
                    continue
            except ValueError:
                continue

            date_str = (
                row.get("effective_date_time")
                or row.get("effectiveDateTime")
                or row.get("timestamp")
                or ""
            )
            source = row.get("source_type") or "unknown"
            unit = (row.get("unit") or "").strip()
            timestamp = parse_timestamp(date_str)

            measurements.append(
                {
                    "measurementId": measurement_id,
                    "weight": weight,
                    "unit": unit,
                    "timestamp": timestamp,
                    "source": source,
                }
            )

    # Sort chronologically
    measurements.sort(key=lambda m: m["timestamp"])

    target_index = next(
        (i for i, m in enumerate(measurements) if m["measurementId"] == target_measurement_id),
        None,
    )

    print(f"\n=== Quality Component Breakdown ===")
    print(f"Target measurement: {target_measurement_id}")
    print(f"Processing first {target_index + 1} measurements...\n")

    # Initialize
    state_store = InMemoryStore()
    config = get_default_config()
    processor = Processor(user_id, state_store, config)

    # Process measurements up to and including target
    for i, m_data in enumerate(measurements[: target_index + 1]):
        print(f"\n[{i}] Processing {m_data['measurementId'][:8]}...")
        print(f"   Timestamp: {m_data['timestamp'].isoformat()}")
        print(f"   Weight: {m_data['weight']} {m_data['unit']}")

        result = processor.process_measurement(
            measurement_id=m_data["measurementId"],
            weight=m_data["weight"],
            unit=m_data["unit"],
            timestamp=m_data["timestamp"],
            source=m_data["source"],
        )

        print(f"   Accepted: {result.accepted}")
        print(f"   Quality Score: {result.quality_score:.15f}")

        # For the target measurement, get detailed breakdown
        if m_data["measurementId"] == target_measurement_id:
            print(f"\n🎯 DETAILED BREAKDOWN FOR DIVERGENT MEASUREMENT:")

            if result.quality_details:
                print(f"\n   Quality Components:")

                components = result.quality_details.get("component_scores", {})
                weights = config["quality_scoring"]["components"]

                for name, score in components.items():
                    weight = weights.get(name, {}).get("weight", 0)
                    contribution = score * weight
                    print(f"     {name}:")
                    print(f"       Raw score: {score:.15f}")
                    print(f"       Weight: {weight}")
                    print(f"       Contribution: {contribution:.15f}")

                print(f"\n   Final Quality Score: {result.quality_score:.15f}")
                print(f"   Threshold: 0.5")
                print(f"   Decision: {'ACCEPT' if result.accepted else 'REJECT'}")

                # Check if detailed metrics available
                if "metrics" in result.quality_details:
                    print(f"\n   Additional Metrics:")
                    print(f"   {json.dumps(result.quality_details['metrics'], indent=2)}")
            else:
                print(f"   ⚠️  No quality_details available in result")

            # Get Kalman state
            state = state_store.get_state(user_id)
            if state and "kalman_filter" in state:
                kf = state["kalman_filter"]
                print(f"\n   Kalman State:")
                print(f"     Weight: {kf['state'][0]:.15f}")
                print(f"     Velocity: {kf['state'][1]:.15f}")
                cov = kf["covariance"]
                print(f"     Covariance: [[{', '.join(f'{v:.6f}' for v in cov[0])}],")
                print(f"                  [{', '.join(f'{v:.6f}' for v in cov[1])}]]")

            # Write detailed results
            breakdown = {
                "measurementId": m_data["measurementId"],
                "measurementIndex": i,
                "timestamp": m_data["timestamp"].isoformat(),
                "weight": m_data["weight"],
                "unit": m_data["unit"],
                "source": m_data["source"],
                "qualityScore": result.quality_score,
                "accepted": result.accepted,
                "qualityDetails": result.quality_details,
                "kalmanState": (
                    {
                        "weight": float(state["kalman_filter"]["state"][0]),
                        "velocity": float(state["kalman_filter"]["state"][1]),
                        "covariance": state["kalman_filter"]["covariance"],
                    }
                    if state and "kalman_filter" in state
                    else None
                ),
            }

            with open("py_quality_breakdown.json", "w") as f:
                json.dump(breakdown, f, indent=2)

            print(f"\n✅ Detailed breakdown written to py_quality_breakdown.json")


if __name__ == "__main__":
    main()

File: test_quality_breakdown.ts
#!/usr/bin/env bun
/**
 * Quality Component Breakdown for Divergent Measurement
 *
 * Processes the first divergent measurement (#1) and extracts
 * detailed breakdown of each quality component
 */

import { readFileSync } from "node:fs";
import { parse } from "csv-parse/sync";
import { InMemoryStore } from "./typescript_lib/src/index";
import { processMeasurement } from "./typescript_lib/src/weight-processor-lib/core/processing/processor";
import type { MeasurementInput } from "./services/weight_processor_service";

function parseTimestamp(dateStr: string): Date {
  if (!dateStr) return new Date();
  try {
    if (dateStr.includes("T")) {
      return new Date(dateStr.replace("Z", "+00:00"));
    } else if (dateStr.includes(" ")) {
      return new Date(dateStr.replace(" ", "T") + "Z");
    } else {
      return new Date(dateStr + "T00:00:00Z");
    }
  } catch {
    return new Date();
  }
}

function getDefaultConfig(): any {
  return {
    database: { backend: "memory" },
    kalman: {
      initial_variance: 0.364,
      transition_covariance_weight: 0.018,
      transition_covariance_trend: 0.00015,
      observation_covariance: 3.49,
    },
    quality_scoring: {
      threshold: 0.5,
      components: {
        kalman_fit: { weight: 0.3, enabled: true },
        temporal_consistency: { weight: 0.25, enabled: true },
        anomaly_detection: { weight: 0.25, enabled: true },
        source_reliability: { weight: 0.1, enabled: true },
        trend_alignment: { weight: 0.1, enabled: true },
      },
    },
    processing: { enable_validation: true, enable_quality_scoring: true },
    reset: { time_gap_days: 30, weight_change_threshold_kg: 10 },
    snapshot: { interval_hours: 24, periodic_enabled: true },
    adaptive_noise: { enabled: true },
    replay: { buffered_replay_enabled: false, buffer_hours: 24, max_buffer_measurements: 100 },
  };
}

interface CsvRow {
  id?: string;
  measurement_id?: string;
  user_id: string;
  value_quantity?: string;
  weight?: string;
  unit: string;
  timestamp?: string;
  effective_date_time?: string;
  effectiveDateTime?: string;
  source_type: string;
  [key: string]: any;
}

async function main() {
  const csvPath = "test_user.csv";
  const userId = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7";
  const targetMeasurementId = "0bb4ca6c-d123-4461-8cae-a40297230843"; // Divergent measurement #1

  // Load CSV
  const content = readFileSync(csvPath, "utf-8");
  const records = parse(content, {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  }) as CsvRow[];

  // Parse measurements
  const measurements: MeasurementInput[] = [];
  for (const row of records) {
    const measurementId = row.id || row.measurement_id;
    if (!measurementId || row.user_id !== userId) continue;

    const weightStr = (row.value_quantity || row.weight || "").trim();
    if (!weightStr || weightStr.toUpperCase() === "NULL") continue;

    const weight = parseFloat(weightStr);
    if (weight <= 0 || weight > 1000 || isNaN(weight) || !isFinite(weight)) continue;

    const dateStr = row.effective_date_time || row.effectiveDateTime || row.timestamp || "";
    const source = row.source_type || "unknown";
    const unit = (row.unit || "").trim();
    const timestamp = dateStr ? parseTimestamp(dateStr) : new Date();

    measurements.push({ measurementId, weight, unit, timestamp, source });
  }

  // Sort chronologically
  measurements.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());

  console.log(`\n=== Quality Component Breakdown ===`);
  console.log(`Target measurement: ${targetMeasurementId}`);
  console.log(`Processing first ${measurements.findIndex(m => m.measurementId === targetMeasurementId) + 1} measurements...\n`);

  // Initialize
  const stateStore = new InMemoryStore();
  const config = getDefaultConfig();

  // Process measurements up to and including target
  const targetIndex = measurements.findIndex(m => m.measurementId === targetMeasurementId);

  for (let i = 0; i <= targetIndex; i++) {
    const measurement = measurements[i];

    console.log(`\n[${i}] Processing ${measurement.measurementId.substring(0, 8)}...`);
    console.log(`   Timestamp: ${measurement.timestamp.toISOString()}`);
    console.log(`   Weight: ${measurement.weight} ${measurement.unit}`);

    const result = await processMeasurement(
      userId,
      measurement.weight,
      measurement.timestamp,
      measurement.source,
      config,
      measurement.unit,
      stateStore,
      null // user height
    );

    console.log(`   Accepted: ${result.accepted}`);
    console.log(`   Quality Score: ${result.quality_score.toFixed(15)}`);

    // For the target measurement, get detailed breakdown
    if (measurement.measurementId === targetMeasurementId) {
      console.log(`\n🎯 DETAILED BREAKDOWN FOR DIVERGENT MEASUREMENT:`);

      if (result.quality_details) {
        console.log(`\n   Quality Components:`);

        const components = result.quality_details.component_scores || {};
        const weights = config.quality_scoring.components;

        for (const [name, score] of Object.entries(components)) {
          const weight = weights[name]?.weight || 0;
          const contribution = (score as number) * weight;
          console.log(`     ${name}:`);
          console.log(`       Raw score: ${(score as number).toFixed(15)}`);
          console.log(`       Weight: ${weight}`);
          console.log(`       Contribution: ${contribution.toFixed(15)}`);
        }

        console.log(`\n   Final Quality Score: ${result.quality_score.toFixed(15)}`);
        console.log(`   Threshold: 0.5`);
        console.log(`   Decision: ${result.accepted ? "ACCEPT" : "REJECT"}`);

        // Check if detailed metrics available
        if (result.quality_details.metrics) {
          console.log(`\n   Additional Metrics:`);
          console.log(`   ${JSON.stringify(result.quality_details.metrics, null, 2)}`);
        }
      } else {
        console.log(`   ⚠️  No quality_details available in result`);
      }

      // Get Kalman state
      const state = await stateStore.getState(userId);
      if (state?.kalman_filter) {
        console.log(`\n   Kalman State:`);
        console.log(`     Weight: ${state.kalman_filter.state[0].toFixed(15)}`);
        console.log(`     Velocity: ${state.kalman_filter.state[1].toFixed(15)}`);
        console.log(`     Covariance: [[${state.kalman_filter.covariance[0].map(v => v.toFixed(6)).join(", ")}],`);
        console.log(`                  [${state.kalman_filter.covariance[1].map(v => v.toFixed(6)).join(", ")}]]`);
      }

      // Write detailed results
      const breakdown = {
        measurementId: measurement.measurementId,
        measurementIndex: i,
        timestamp: measurement.timestamp.toISOString(),
        weight: measurement.weight,
        unit: measurement.unit,
        source: measurement.source,
        qualityScore: result.quality_score,
        accepted: result.accepted,
        qualityDetails: result.quality_details,
        kalmanState: state?.kalman_filter ? {
          weight: state.kalman_filter.state[0],
          velocity: state.kalman_filter.state[1],
          covariance: state.kalman_filter.covariance,
        } : null,
      };

      Bun.write("ts_quality_breakdown.json", JSON.stringify(breakdown, null, 2));
      console.log(`\n✅ Detailed breakdown written to ts_quality_breakdown.json`);
    }
  }
}

main().catch(console.error);
